sbt-plugin/src/main/scala/scala/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -409,54 +475,41 @@ object ScalaJSPluginInternal {
           true
       },
 
-      scalaJSRequestsDOM :=
-        requiresDOM.?.value.getOrElse(scalaJSExecClasspath.value.requiresDOM),
-
-      // Default jsEnv
-      jsEnv <<= Def.taskDyn {
-        scalaJSStage.value match {
-          case Stage.PreLink =>
-            Def.task {
-              preLinkJSEnv.?.value.getOrElse {
-                new RhinoJSEnv(scalaJSSemantics.value,
-                    withDOM = scalaJSRequestsDOM.value)
-              }
-            }
-          case Stage.FastOpt | Stage.FullOpt =>
-            Def.task(scalaJSDefaultPostLinkJSEnv.value)
-        }
-      },
+      scalaJSRequestsDOM := requiresDOM.?.value
+          .getOrElse(jsDependencyManifests.value.data.exists(_.requiresDOM)),
 
-      // Wire jsEnv and sources for other stages
-      scalaJSDefaultPostLinkJSEnv := postLinkJSEnv.?.value.getOrElse {
-        if (scalaJSRequestsDOM.value)
+      resolvedJSEnv := jsEnv.?.value.getOrElse {
+        if (scalaJSUseRhino.value)
+          new RhinoJSEnv(withDOM = scalaJSRequestsDOM.value)
+        else if (scalaJSRequestsDOM.value)
           new PhantomJSEnv(jettyClassLoader = scalaJSPhantomJSClassLoader.value)
         else
           new NodeJSEnv
       },
 
-      scalaJSExecClasspath <<= Def.taskDyn {
-        scalaJSStage.value match {
-          case Stage.PreLink =>
-            Def.task { scalaJSLinkingUnitClasspath.value }
-          case Stage.FastOpt =>
-            Def.task { fastOptJS.value.get(scalaJSCompleteClasspath).get }
-          case Stage.FullOpt =>
-            Def.task { fullOptJS.value.get(scalaJSCompleteClasspath).get }
+      scalaJSLibEnv <<= Def.taskDyn {
+        val log = streams.value.log
+        val libs = resolvedJSDependencies.value.data
+        resolvedJSEnv.value match {
+          case env: LinkingUnitJSEnv =>
+            log.debug("Loading JSEnv with LinkingUnit")
+            Def.task(env.loadLibs(libs, scalaJSLinkingUnit.value))
+          case env =>
+            log.debug("Loading JSEnv with linked file")
+            Def.task(env.loadLibs(libs, scalaJSLinkedFile.value))
```

Comment:
> However, in view that Scala.js code should just become a normal library, another solution might be to have loadLibs to return a JSEnv again which just "magically" loads the relevant libs. This will significantly reduce the number of classes exposed in the API as well.

Yes, that sounds like a good idea.